plat/common/platform_bus.c

```diff
@@ -82,104 +93,96 @@ static inline struct pf_driver *pf_find_driver(struct pf_device_id *id)
 }
 
+static inline int pf_driver_probe_device(struct pf_driver *drv,
+					struct pf_device *dev)
+{
+	int ret;
+
+	UK_ASSERT(drv != NULL && dev != NULL);
+	UK_ASSERT(drv->probe != NULL);
+
+	uk_pr_info("pf_driver_probe_device devid=%d\n", dev->id.device_id);
+
+	ret = drv->probe(dev);
+	if (ret < 0) {
+		uk_pr_err("Platform Failed to probe device driver\n");
+
+		return ret;
+	}
+
+	return 0;
+}
+
 static int pf_probe(void)
 {
-	struct pf_device_id devid;
 	struct pf_driver *drv;
-	int i;
-	int end_offset = -1;
+	int idx = 0;
 	int ret = -ENODEV;
-	const fdt32_t *prop;
-	int type, hwirq, prop_len;
-	__u64 reg_base;
-	__phys_addr dev_base;
-	int dev_irq;
+	struct pf_device *dev;
+	int fdt_pf = -1;
 
 	uk_pr_info("Probe PF\n");
 
-	/* We only support virtio_mmio as a platform device here.
-	 * A loop here is needed for finding drivers if more devices
-	 */
-	devid.device_id = VIRTIO_MMIO_ID;
+	/* Search all the platform bus devices provided by fdt */
+	do {
+		fdt_pf = fdt_node_offset_idx_by_compatible_list(_libkvmplat_cfg.dtb,
+						fdt_pf, pf_device_compatible_list, &idx);
+		if (fdt_pf < 0) {
+			uk_pr_info("End of searching platform devices\n");
+			break;
+		}
 
-	drv = pf_find_driver(&devid);
-	if (!drv) {
-		uk_pr_info("<no driver>\n");
-		return -ENODEV;
-	}
+		/* Alloc dev */
+		dev = (struct pf_device *) uk_calloc(pfh.a, 1, sizeof(*dev));
+		if (!dev) {
+			uk_pr_err("Platform : Failed to initialize: Out of memory!\n");
+			return -ENOMEM;
+		}
 
-	uk_pr_info("driver %p\n", drv);
-
-	/* qemu creates virtio devices in reverse order */
-	for (i = 0; i < UK_MAX_VIRTIO_MMIO_DEVICE; i++) {
-		end_offset = fdt_node_offset_by_compatible_list(fdt_start,
-							end_offset,
-							pf_device_list);
-		if (end_offset == -FDT_ERR_NOTFOUND) {
-			uk_pr_info("device not found in fdt\n");
-			goto error_exit;
-		} else {
-			prop = fdt_getprop(fdt_start, end_offset, "interrupts", &prop_len);
-			if (!prop) {
-				uk_pr_err("irq of device not found in fdt\n");
-				goto error_exit;
-			}
+		dev->fdt_offset = fdt_pf;
 
-			type = fdt32_to_cpu(prop[0]);
-			hwirq = fdt32_to_cpu(prop[1]);
+		/* Find drv with compatible-id match table */
+		drv = pf_find_driver(pf_device_compatible_list[idx]);
+		if (!drv) {
+			uk_free(pfh.a, dev);
+			continue;
+		}
 
-			prop = fdt_getprop(fdt_start, end_offset, "reg", &prop_len);
-			if (!prop) {
-				uk_pr_err("reg of device not found in fdt\n");
-				goto error_exit;
-			}
+		dev->id = *(struct pf_device_id *)drv->device_ids;
+		uk_pr_info("driver %p\n", drv);
 
-			/* only care about base addr, ignore the size */
-			reg_base = fdt32_to_cpu(prop[0]);
-			reg_base = reg_base << 32 | fdt32_to_cpu(prop[1]);
+		ret = pf_driver_probe_device(drv, dev);
+		if (ret < 0) {
+			uk_free(pfh.a, dev);
+			continue;
 		}
 
-		dev_base = reg_base;
-		dev_irq = gic_irq_translate(type, hwirq);
-
-		ret = pf_driver_add_device(drv, &devid, dev_base, dev_irq);
-	}
+		ret = pf_driver_add_device(drv, dev);
+		if (ret < 0) {
```

Comment:
I think the condition here should be `ret < 0  && ret != -ENODEV` in order to avoid confusing error log.

Reply:
Yes，it has been changed in https://github.com/unikraft/unikraft/commit/8edd8dfa0d51d9a1cc2759ed28f11ff35bd656f8

Reply:
That commit changes a similar condition in pf_driver_add_device(), at line 107, but not this one.

Reply:
The problem is caused by a difference between https://github.com/unikraft/unikraft/commit/8edd8dfa0d51d9a1cc2759ed28f11ff35bd656f8 and this patch: https://patchwork.unikraft.org/patch/738335/
In  https://github.com/unikraft/unikraft/commit/8edd8dfa0d51d9a1cc2759ed28f11ff35bd656f, line 179 is not removed.

Reply:
Yes, `uk_free` should remain even if `ret == -ENODEV`. Only removing the error message in that place is enough, so everything looks good now.